File: OpenStreetMap/OSMiOS/OSMiOS/AutocompleteTextField.m
//
//  AutocompleteTextField.m
//  Go Map!!
//
//  Created by Bryce on 2/3/13.
//  Copyright (c) 2013 Bryce. All rights reserved.
//

#import <QuartzCore/QuartzCore.h>

#import "AutocompleteTextField.h"


@interface AutocompleteTextFieldDelegate : NSObject<UITextFieldDelegate>
@property (strong,nonatomic)	id<UITextFieldDelegate>		realDelegate;
@property (weak,nonatomic)		AutocompleteTextField	*	owner;
@end

@implementation AutocompleteTextFieldDelegate

#if 0
- (BOOL)respondsToSelector:(SEL)aSelector
{
	if ( aSelector == @selector(textFieldShouldBeginEditing:) )
		return [_realDelegate respondsToSelector:aSelector];
	if ( aSelector == @selector(textFieldDidBeginEditing:) )
		return [_realDelegate respondsToSelector:aSelector];
	if ( aSelector == @selector(textFieldShouldEndEditing:) )
		return [_realDelegate respondsToSelector:aSelector];
	if ( aSelector == @selector(textFieldDidEndEditing:) )
		return YES || [_realDelegate respondsToSelector:aSelector];
	if ( aSelector == @selector(textFieldShouldClear:) )
		return YES || [_realDelegate respondsToSelector:aSelector];
	if ( aSelector == @selector(textFieldShouldReturn:) )
		return [_realDelegate respondsToSelector:aSelector];
	if ( aSelector == @selector(textField:shouldChangeCharactersInRange:replacementString:) )
		return YES || [_realDelegate respondsToSelector:aSelector];

	return NO;
}
#endif

- (BOOL)textFieldShouldBeginEditing:(UITextField *)textField
{
	if ( [_realDelegate respondsToSelector:@selector(textFieldShouldBeginEditing:)])
		return [_realDelegate textFieldShouldBeginEditing:textField];
	return YES;
}
- (void)textFieldDidBeginEditing:(UITextField *)textField
{
	if ( [_realDelegate respondsToSelector:@selector(textFieldDidBeginEditing:)])
		[_realDelegate textFieldDidBeginEditing:textField];
}
- (BOOL)textFieldShouldEndEditing:(UITextField *)textField
{
	if ( [_realDelegate respondsToSelector:@selector(textFieldShouldEndEditing:)])
		return [_realDelegate textFieldShouldEndEditing:textField];
	return YES;
}
- (void)textFieldDidEndEditing:(UITextField *)textField
{
	[_owner clearFilteredCompletionsInternal];

	if ( [_realDelegate respondsToSelector:@selector(textFieldDidEndEditing:)])
		[_realDelegate textFieldDidEndEditing:textField];
}
- (BOOL)textFieldShouldClear:(UITextField *)textField
{
#if 1
	BOOL result = [_realDelegate respondsToSelector:@selector(textField:textFieldShouldClear:)] ? [_realDelegate textFieldShouldClear:textField] : YES;
	if ( result ) {
		[self.owner performSelector:@selector(updateAutocompleteForString:) withObject:@""];
	}
	return result;
#else
	[_owner performSelector:@selector(updateAutocomplete) withObject:nil afterDelay:0.0];

	if ( [_realDelegate respondsToSelector:@selector(textFieldShouldClear:)])
		return [_realDelegate textFieldShouldClear:textField];
	return YES;
#endif
}
- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
	if ( [_realDelegate respondsToSelector:@selector(textFieldShouldReturn:)])
		return [_realDelegate textFieldShouldReturn:textField];
	return YES;
}
- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
#if 1
	BOOL result = [_realDelegate respondsToSelector:@selector(textField:shouldChangeCharactersInRange:replacementString:)] ? [_realDelegate textField:textField shouldChangeCharactersInRange:range replacementString:string] : YES;
	if ( result ) {
		NSString * newString = [self.owner.text stringByReplacingCharactersInRange:range withString:string];
		[self.owner performSelector:@selector(updateAutocompleteForString:) withObject:newString];
	}
	return result;
#else
	[_owner performSelector:@selector(updateAutocomplete) withObject:nil afterDelay:0.0];

	if ( [_realDelegate respondsToSelector:@selector(textField:shouldChangeCharactersInRange:replacementString:)])
		return [_realDelegate textField:textField shouldChangeCharactersInRange:range replacementString:string];
	return YES;
#endif
}

@end



@implementation AutocompleteTextField
@synthesize completions = _allCompletions;

static const CGFloat GradientHeight = 20.0;

- (id)initWithCoder:(NSCoder *)coder
{
	self = [super initWithCoder:coder];

	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillChange:) name:UIKeyboardWillChangeFrameNotification object:nil];

	_myDelegate = [AutocompleteTextFieldDelegate new];
	_myDelegate.owner = self;
	super.delegate = _myDelegate;
	
	return self;
}

- (id)initWithFrame:(CGRect)frame
{
	self = [super initWithFrame:frame];
	if (self) {
		assert(NO);
	}
	return self;
}

-(void)setDelegate:(id<UITextFieldDelegate>)delegate
{
	_myDelegate.realDelegate = delegate;
	super.delegate = _myDelegate;
}

-(id<UITextFieldDelegate>)delegate
{
	return _myDelegate.realDelegate;
}

-(void)clearFilteredCompletionsInternal
{
	_filteredCompletions = nil;
	[self updateCompletionTableView];
}

-(void)setCompletions:(NSArray *)completions
{
	_allCompletions = completions;
	if ( self.delegate != _myDelegate ) {
		_myDelegate.realDelegate = self.delegate;
		super.delegate = _myDelegate;
	}
}
-(NSArray *)completions
{
	return _allCompletions;
}


- (CGSize)keyboardSizeFromNotification:(NSNotification *)notification
{
	NSDictionary *userInfo = [notification userInfo];
	CGRect rect = [[userInfo objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
	rect = [self.superview convertRect:rect fromView:nil];
	return rect.size;
}

- (void) keyboardWillShow:(NSNotification *)nsNotification
{
	_keyboardSize = [self keyboardSizeFromNotification:nsNotification];

	if ( self.editing && _filteredCompletions.count ) {
		[self updateAutocomplete];
	}
}
- (void) keyboardWillChange:(NSNotification *)nsNotification
{
	_keyboardSize = [self keyboardSizeFromNotification:nsNotification];

	if ( _completionTableView ) {
		CGRect rect = [self frameForCompletionTableView];
		_completionTableView.frame = rect;

		CGRect rcGradient = rect;
		rcGradient.size.height = GradientHeight;

		[CATransaction begin];
		[CATransaction setAnimationDuration:0.0];
		_gradientLayer.frame = rcGradient;
		[CATransaction commit];
	}
	if ( self.editing && _filteredCompletions.count ) {
		[self updateAutocomplete];
	}
}

-(CGRect)frameForCompletionTableView
{
	BOOL iOS7 = NO;
	UITableViewCell * cell = (id)[self.superview superview];
	UITableView * tableView = (id)[cell superview];
	if ( [tableView isKindOfClass:[UITableViewCell class]] ) {
		// ios 7
		tableView = (id)tableView.superview.superview;
		cell = (id)cell.superview;
		iOS7 = YES;
	}
	UIScrollView * scrollView = (id)tableView.superview;
	UIView * view = scrollView.superview;

	CGRect rect;
	rect.origin.x = cell.frame.origin.x;
	rect.origin.y = cell.frame.origin.y + cell.frame.size.height;
	rect.size.width = cell.frame.size.width;
	rect.size.height = view.frame.size.height - _keyboardSize.height - cell.frame.size.height;
	return rect;
}

- (void)updateCompletionTableView
{
	if ( _filteredCompletions.count ) {
		if ( _completionTableView == nil ) {

			UITableViewCell * cell = (id)self.superview.superview;
			UITableView * tableView = (id)cell.superview;
			BOOL iOS7 = NO;
			if ( [tableView isKindOfClass:[UITableViewCell class]] ) {
				// iOS 7
				iOS7 = YES;
				tableView = (id)tableView.superview.superview;
				cell = (id)cell.superview;
			}
			assert( [tableView isKindOfClass:[UITableView class]] );

			// add completion table to tableview
			CGRect rect = [self frameForCompletionTableView];
			_completionTableView = [[UITableView alloc] initWithFrame:rect style:UITableViewStylePlain];
			_completionTableView.backgroundColor = [UIColor colorWithWhite:0.88 alpha:1.0];
			_completionTableView.separatorColor = [UIColor colorWithWhite:0.7 alpha:1.0];
			_completionTableView.dataSource = self;
			_completionTableView.delegate = self;
			[tableView addSubview:_completionTableView];

			_gradientLayer = [CAGradientLayer layer];
			_gradientLayer.colors = @[
						(id)[UIColor colorWithWhite:0.0 alpha:0.6].CGColor,
						(id)[UIColor colorWithWhite:0.0 alpha:0.0].CGColor ];
			CGRect rcGradient = rect;
			rcGradient.size.height = GradientHeight;
			_gradientLayer.frame = rcGradient;
			[tableView.layer addSublayer:_gradientLayer];

			// scroll cell to top
			CGRect cellFrame = cell.frame;
			if ( iOS7 ) {
				cellFrame.origin.y -= 45 + 20;
			}
			[tableView setContentOffset:CGPointMake(0,cellFrame.origin.y) animated:YES];
			tableView.scrollEnabled = NO;
		}
		[_completionTableView reloadData];

	} else {
		[_completionTableView removeFromSuperview];
		_completionTableView = nil;

		[_gradientLayer removeFromSuperlayer];
		_gradientLayer = nil;

		UITableViewCell * cell = (id)[self.superview superview];
		UITableView * tableView = (id)[cell superview];
		if ( [tableView isKindOfClass:[UITableViewCell class]] ) {
			// ios 7
			tableView = (id)tableView.superview.superview;
		}
		NSIndexPath * cellIndexPath = [tableView indexPathForCell:cell];
		[tableView scrollToRowAtIndexPath:cellIndexPath atScrollPosition:UITableViewScrollPositionMiddle animated:YES];

		tableView.scrollEnabled = YES;
	}
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
	self.text = _filteredCompletions[ indexPath.row ];

	[self sendActionsForControlEvents:UIControlEventEditingChanged];
	// [[NSNotificationCenter defaultCenter] postNotificationName:UITextFieldTextDidChangeNotification object:self userInfo:nil];

	// hide completion table view
	_filteredCompletions = nil;
	[self updateCompletionTableView];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
	return _filteredCompletions.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString * CellIdentifier = @"Cell";
	UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
	if ( cell == nil ) {
		cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
		cell.textLabel.font = [UIFont boldSystemFontOfSize:16];
	}
	cell.textLabel.text = [_filteredCompletions objectAtIndex:indexPath.row];
    return cell;
}

- (void)updateAutocompleteForString:(NSString *)text
{
	// filter completion list by current text
	if ( text.length ) {
		_filteredCompletions = [_allCompletions filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(NSString * object, NSDictionary *bindings) {
			return [object rangeOfString:text options:NSCaseInsensitiveSearch].location == 0;
		}]];
	} else {
		_filteredCompletions = _allCompletions;
	}
	// sort alphabetically
	_filteredCompletions = [_filteredCompletions sortedArrayUsingComparator:^NSComparisonResult(NSString * s1, NSString * s2) {
		return [s1 compare:s2 options:NSCaseInsensitiveSearch];
	}];
	[self updateCompletionTableView];
}

- (void)updateAutocomplete
{
	[self updateAutocompleteForString:self.text];
}

@end

File: src/iOS/CustomViews/AutocompleteTextField.swift
//
//  AutocompleteTextField.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 2/3/13.
//  Copyright (c) 2013 Bryce Cogswell. All rights reserved.
//

import UIKit

class AutocompleteTextField: UITextField, UITableViewDataSource, UITableViewDelegate {
	// this needs to be shared, because sometimes we'll create a new autocomplete text field when the keyboard is already showing,
	// so it never gets a chance to retrieve the size:
	private static var s_keyboardFrame = CGRect.zero
	private static let GradientHeight: CGFloat = 20.0

	var completionTableView: UITableView?
	var origCellOffset: CGFloat = 0.0
	var filteredCompletions: [String] = []
	var gradientLayer: CAGradientLayer?
	var didSelectAutocomplete: (() -> Void)?

	required init?(coder: NSCoder) {
		super.init(coder: coder)

		NotificationCenter.default.addObserver(
			self,
			selector: #selector(keyboardWillShow(_:)),
			name: UIResponder.keyboardWillShowNotification,
			object: nil)
		NotificationCenter.default.addObserver(
			self,
			selector: #selector(keyboardWillChange(_:)),
			name: UIResponder.keyboardWillChangeFrameNotification,
			object: nil)

		addTarget(self, action: #selector(editingChanged), for: .editingChanged)
		addTarget(self, action: #selector(editingEnded), for: .editingDidEnd)
	}

	deinit {
		NotificationCenter.default.removeObserver(self)
	}

	override init(frame: CGRect) {
		super.init(frame: frame)
		assertionFailure() // not supported
	}

	func clearFilteredCompletionsInternal() {
		filteredCompletions = []
		updateCompletionTableView()
	}

	private var allStrings: [String] = []

	var autocompleteStrings: [String] {
		get {
			return allStrings
		}
		set(strings) {
			allStrings = strings
		}
	}

	func updateAutocomplete(for text: String) {
		let text = text.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
		// filter completion list by current text
		if text.count != 0 {
			filteredCompletions = allStrings.filter({ $0.range(of: text, options: .caseInsensitive) != nil })
		} else {
			filteredCompletions = allStrings
		}
		// sort by prefix, then alphabetically
		filteredCompletions = filteredCompletions.sorted(by: { s1, s2 in
			let p1 = s1.hasPrefix(text)
			let p2 = s2.hasPrefix(text)
			if p1 == p2 {
				// both have, or don't have, a matching prefix, so use alphabetical
				return s1.caseInsensitiveCompare(s2) == .orderedAscending
			} else {
				// matching prefix first
				return p1
			}
		})
		updateCompletionTableView()
	}

	func keyboardFrame(from notification: Notification) -> CGRect {
		let userInfo = notification.userInfo
		let rect = (userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as! NSValue).cgRectValue
		return rect
	}

	@objc func keyboardWillShow(_ nsNotification: Notification) {
		AutocompleteTextField.s_keyboardFrame = keyboardFrame(from: nsNotification)

		if isEditing, filteredCompletions.count != 0 {
			updateAutocomplete()
		}
	}

	// keyboard size can change if switching languages inside keyboard, etc.
	@objc func keyboardWillChange(_ nsNotification: Notification) {
		AutocompleteTextField.s_keyboardFrame = keyboardFrame(from: nsNotification)

		if completionTableView != nil {
			let rect = frameForCompletionTableView()
			completionTableView?.frame = rect

			var rcGradient = rect
			rcGradient.size.height = AutocompleteTextField.GradientHeight

			CATransaction.begin()
			CATransaction.setAnimationDuration(0.0)
			gradientLayer?.frame = rcGradient
			CATransaction.commit()
		}
		if isEditing, filteredCompletions.count != 0 {
			updateAutocomplete()
		}
	}

	func frameForCompletionTableView() -> CGRect {
		guard let cell = superviewOfType(UITableViewCell.self),
		      let tableView = cell.superviewOfType(UITableView.self)
		else {
			return .zero
		}

		let cellRC = cell.convert(cell.bounds, to: tableView)
		var rect = CGRect.zero
		rect.origin.x = 0
		rect.origin.y = cellRC.origin.y + cellRC.size.height
		rect.size.width = tableView.frame.size.width
		if AutocompleteTextField.s_keyboardFrame.size.height > 0 {
			let keyboardPos = tableView
				.convert(AutocompleteTextField.s_keyboardFrame, from: nil) // keyboard is in screen coordinates
			rect.size.height = keyboardPos.origin.y - rect.origin.y
		} else {
			// no on-screen keyboard (external keyboard or Mac Catalyst)
			rect.size.height = tableView.frame.size.height - cellRC.size.height
		}
		return rect
	}

	func updateCompletionTableView() {
		if filteredCompletions.count != 0 {
			if completionTableView == nil {
				guard
					let cell = superviewOfType(UITableViewCell.self),
					let tableView = cell.superviewOfType(UITableView.self),
					let indexPath = tableView.indexPath(for: cell)
				else { return }
				// scroll cell to top
				tableView.scrollToRow(at: indexPath, at: .top, animated: false)

				if #available(iOS 15, *),
				   ProcessInfo.processInfo.operatingSystemVersion.majorVersion < 26
				{
					// iOS 15 ..< iOS 26, handles scrolling differently and disabling it causes visual glitches
					tableView.isScrollEnabled = true
				} else {
					tableView.isScrollEnabled = false
				}

				// cell doesn't always scroll to the same place, so give it a moment before we add the completion table
				DispatchQueue.main.async(execute: {
					// add completion table to tableview
					let backgroundColor: UIColor
					if #available(iOS 13.0, *) {
						backgroundColor = UIColor.systemBackground
					} else {
						backgroundColor = UIColor(white: 0.88, alpha: 1.0)
					}

					let rect = self.frameForCompletionTableView()
					let view = UITableView(frame: rect, style: .plain)
					view.backgroundColor = backgroundColor
					view.separatorColor = UIColor(white: 0.7, alpha: 1.0)
					view.dataSource = self
					view.delegate = self
					tableView.addSubview(view)
					self.completionTableView = view

					let layer = CAGradientLayer()
					layer.colors = [
						UIColor(white: 0.0, alpha: 0.6).cgColor,
						UIColor(white: 0.0, alpha: 0.0).cgColor
					]
					var rcGradient = rect
					rcGradient.size.height = AutocompleteTextField.GradientHeight
					layer.frame = rcGradient
					tableView.layer.addSublayer(layer)
					self.gradientLayer = layer
				})
			}
			completionTableView?.reloadData()
		} else {
			completionTableView?.removeFromSuperview()
			completionTableView = nil

			gradientLayer?.removeFromSuperlayer()
			gradientLayer = nil

			if let cell = superviewOfType(UITableViewCell.self),
			   let tableView = cell.superviewOfType(UITableView.self)
			{
				if let cellIndexPath = tableView.indexPath(for: cell) {
					tableView.scrollToRow(at: cellIndexPath, at: .middle, animated: true)
				}
				tableView.isScrollEnabled = true
			}
		}
	}

	func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		guard indexPath.row < filteredCompletions.count else { return }

		text = filteredCompletions[indexPath.row]

		sendActions(for: .editingChanged)

		didSelectAutocomplete?()

		// hide completion table view
		filteredCompletions = []
		updateCompletionTableView()
	}

	func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return filteredCompletions.count
	}

	func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let cellIdentifier = "Cell"
		let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
			?? UITableViewCell(style: .default, reuseIdentifier: cellIdentifier)
		cell.textLabel?.font = UIFont.preferredFont(forTextStyle: .body)
		// shouldn't need this check but we got an out-of-range crash report
		if indexPath.row < filteredCompletions.count {
			cell.textLabel?.text = filteredCompletions[indexPath.row]
		} else {
			cell.textLabel?.text = ""
		}
		return cell
	}

	func updateAutocomplete() {
		updateAutocomplete(for: text ?? "")
	}

	@objc private func editingChanged() {
		updateAutocomplete()
	}

	@objc private func editingEnded() {
		clearFilteredCompletionsInternal()
	}

	override func paste(_ sender: Any?) {
		// Check whether they are pasting a set of tags
		if let pb = UIPasteboard.general.string,
		   let tags = OsmTags.tagsForString(pb)
		{
			// try to find an ancestor that we can notify
			var view: UIView? = self
			while view != nil {
				if let cell = view as? KeyValueTableCell {
					cell.keyValueCellOwner?.pasteTags(tags)
					return
				}
				view = view?.superview
			}
		}

		// Do a regular paste
		super.paste(sender)
	}
}
